### PR TITLE
[ALLUXIO-2546] Improve fault tolerance + HDFS API quality of life

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-2546

### CODE REVIEW

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 1]
```diff
@@ -422,15 +422,22 @@ public Path getWorkingDirectory() {
...
+    }
 
     super.initialize(uri, conf);
     LOG.debug("initialize({}, {}). Connecting to Alluxio", uri, conf);
     HadoopUtils.addS3Credentials(conf);
     HadoopUtils.addSwiftCredentials(conf);
     setConf(conf);
-    mAlluxioHeader = getScheme() + "://" + uri.getHost() + ":" + uri.getPort();
+
+    String authority = uri.getHost() + ":" + uri.getPort();
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: do you want to add comments that explain what is going on here?

Comment (calvinjia) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: Does `getAuthority` work better?

Comment (aaudiber) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

Comment (jsimsa) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: have you addressed @calvinjia's comment as well?

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 2]
```diff
@@ -422,15 +422,24 @@ public Path getWorkingDirectory() {
   @SuppressFBWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
   @Override
   public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
-    Preconditions.checkNotNull(uri.getHost(), PreconditionMessage.URI_HOST_NULL);
-    Preconditions.checkNotNull(uri.getPort(), PreconditionMessage.URI_PORT_NULL);
+    if (!Configuration.getBoolean(PropertyKey.ZOOKEEPER_ENABLED)) {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: can you add a comment that explains why is it okay to have empty host / port when using Zookeeper? thanks

Comment (jsimsa) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: nevermind, just saw the documentation you added to file system API

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 3]
```diff
@@ -433,12 +433,8 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
     HadoopUtils.addSwiftCredentials(conf);
     setConf(conf);
 
-    String authority = uri.getHost() + ":" + uri.getPort();
-    if (Configuration.getBoolean(PropertyKey.ZOOKEEPER_ENABLED) && uri.getHost() == null) {
-      // If we are in fault-tolerant mode, the uri host and port may be null. In this case, Hadoop
-      // requires that we use "alluxio-ft:///" for the header.
-      authority = "/";
-    }
+    // HDFS doesn't allow the authority to be ""; it must be "/" instead.
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: replace `""` with empty

Comment (aaudiber) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 2 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: comment added anyway

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 1 above)
Comment (aaudiber) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done done

**Comment by aaudiber** [comment]
@calvinjia comments addressed, PTAL

**Comment by calvinjia** [comment]
LGTM

**Pull request merged**